library/core/src/option.rs

```diff
@@ -985,6 +986,83 @@ impl<T> Option<T> {
     }
 }
 
+impl<T> Option<&T> {
+    /// Converts from `Option<&T>` (or `&Option<&T>`) to `*const T`.
+    ///
+    /// This is the opposite of `<*const T>::as_ref`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_as_ptr)]
+    ///
+    /// let opt = Some(&3);
+    /// let ptr = opt.as_ptr();
+    ///
+    /// assert_eq!(opt, unsafe { ptr.as_ref() });
+    /// ```
+    #[unstable(feature = "option_as_ptr", issue = "none")]
+    #[inline]
+    pub const fn as_ptr(&self) -> *const T {
+        match self {
+            Some(x) => *x,
+            None => ptr::null(),
+        }
+    }
+}
+
+impl<T> Option<&mut T> {
+    /// Converts from `Option<&mut T>` (or `&Option<&mut T>`) to `*const T`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_as_ptr)]
+    ///
+    /// let mut x = 3;
+    /// let opt = Some(&mut x);
+    /// let ptr = opt.as_ptr();
+    ///
+    /// assert_eq!(x, unsafe { *ptr });
+    /// ```
+    #[unstable(feature = "option_as_ptr", issue = "none")]
+    #[rustc_const_unstable(feature = "option_as_ptr", issue = "none")]
```

Comment:
Is there a reason you put `#[rustc_const_unstable]` here, but not on the one above?

Reply:
Because these are over `&mut T`, which is unstable in `const fn`. Whereas the one above is `&T`, which has no issues in `const fn`. Should they both be `#[rustc_const_unstable]` regardless?

Reply:
I'm not sure what the state of that attribute is. In the rustc source code, it says:

https://github.com/rust-lang/rust/blob/8fec6c7bb9f2a549b6f424b3dd1c2c257286bc6d/compiler/rustc_mir/src/transform/check_consts/mod.rs#L113-L116

Which doesn't really give an answer here.

I suppose it's useful to remind us we're also stabilizing the constness of the function at the point when we stabilize these new functions. So it wouldn't hurt to add it to the first one too, just in case.

Reply:
That's a fair point. I'll make this change then.